### DISCUSSION TITLE
브라우저 환경에서의 JavaScript 런타임에 대해 아시나요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#51] [Q&A]
웹 사이트가 돌아가는 일련의 과정(런타임에 영향을 미치는)을 알려주세요.
JS 엔진, 콜스택, 웹 APIs, 콜백큐, 이벤트루프 모두 조금씩이라도 언급해주세요. 

> 관련 내용을 처음 접하신다면
https://www.youtube.com/watch?v=8aGhZQkoFbQ

**Comment by michoball** [comment]
이야기하자만 길지만 간략하게 말해보도록 하겠습니다. 

서버로부터 html, css, 가지고 화면을 만들고나서  js를 받아온 상황에서 시작하겠습니다. 

일반적으로 각 웹 브라우저가 가지고 있는 JS 엔진이 js를 해석하게됩니다. 대표적인 엔진은 v8이 있는데요. 

js를 해석하면서 코드 안에 있는 변수들을 힙에 저장하고, 함수들은 콜 스택으로 들어가게 됩니다. 

js 는 싱글 스레드 언어이기때문에 여러 동작들을 멀티 스레드 언어처럼 동시 동작하는 것처럼 보이게 하면서 동작합니다. 

js 엔진은 콜 스택, 테스크 큐, 마이크로 테스크 큐, RAF 등의 공간을 엄청 빠른 속도로 돌면서 각 장소에 도착했을때 해야하는 업무를 수행합니다. -> 이벤트 루프

js를 읽으면서 만들어진 실행 컨텍스트를 콜스택에 넣습니다. 

콜 스택에  들어온 함수를 우선적으로 스택의 공간이 비워질 때까지 수행합니다.  이때 함수내에 비동기 함수가 있으면 테스크 큐에 넣습니다.  비동시 함수의 콜백함수가 있으면 마이크로 테스크 큐에 넣습니다. 

이후 콜스택이 비워지면 이벤트 루프는 다른 공간을 순회합니다. 

마이크로 테스크 큐에 도착을 하면 여기에 있는 함수들 {비동기 함수의 콜백 함수나 mutation observer(뭔지 잘 모르지만)}을 다 없어질 때까지 수행합니다. -> 수행하면서 새로운 함수가 들어오면 그것도 수행함 다 없어질때까지

그리고 테스크 큐에 도착하면 여기에 있는 비동기 함수 (대체적으로 웹 APIs ) 하나씩 꺼내서 수행합니다. 

중간 중간 1 / 60 초 마다 새로 화면을 그리기 위해서 렌더링을 다시 합니다. 그때 이 렌더링 작업 안에 있는 콜백함수는 RAF에서 수행합니다. 

js 런타임 환경에서는 이러한 과정을 계속 수행하면서 화면을 보여주게 됩니다.

**Comment by javascriptbiggosoo** [comment]
JavaScript 코드가 실행되는 환경을 JavaScript 런타임이라고 합니다
그 구성요소에 대해 말해보겠습니다.
![image](https://github.com/wanted-9th-3team/frontend-study/assets/28939278/e7691cd2-9a67-4660-9d5c-9d468cf9e99c)

### JS ENGINE

js 런타임의 심장. 다음 두가지를 포함합니다.

- **call stack**
    
    코드가 실행되는 곳입니다.
    (여기서 잠깐, excution은 단일 쓰레드로 진행되는데. 즉, 애초 설계가 그렇게 만들어졌기 때문에 js 만으로는 멀티태스킹을 할 수 없다는 점을 짚고갑니다.)
    
- **HEAP**
    
    메모리에서 object들이 따로 저장되는 곳 입니다.
    

### WEB APIs

엔진에 제공되는 API들이 있습니다.
WEB APIs evironment 덕분에 JS환경에서 멀티태스킹이 가능해지는거죠

### CALLBACK QUEUE

이벤트 발생과 같은 특정 시점에 불려올 콜백 함수들이 대기하는 곳입니다.

### EVENT LOOP

콜백을 콜스택으로 넘겨주는 역할을 합니다.
콜스택이 empty일 때마다 이벤트 루프는 콜백큐의 콜백을 가져다 놓습니다.
그래서 그것들이 execute될 수 있덩
비동기 작업이 가능하려면 바로 이 이벤트 루프가 꼭 필요해요.

**Reply by javascriptbiggosoo** [reply]
이벤트 루프 더 자세히 설명하려다 양이 3배쯤 늘어날까봐 멈췄습니다.

**Reply by BaekYuri** [reply]
너무 어려워요 🥹
